### PR TITLE
[SPARK-12896] [WIP] Send only accumulator updates to driver, not TaskMetrics

### PR DESCRIPTION
**Pull request by andrewor14**
This patch builds on top of #10835, so the actual diff should be much smaller. This would complete SPARK-10620, which replaces much of the logic in TaskMetrics with accumulators. This was once part of #10717. It was split into its own patch so it's more reviewable.

**The high level idea is that instead of having the executors send both accumulator updates and TaskMetrics, we should have them send only accumulator updates.** This is because TaskMetrics can be represented as a collection of accumulators (as we have done so in #10835) and we can just reconstruct them on the driver based on the accumulator updates.

While an effort has been made to preserve as much of the public API as possible, there were a few known breaking @DeveloperApi changes that would be very awkward to maintain. I will gather the full list shortly and post it here.

Lastly, this is still WIP because more tests are pending.

### CODE REVIEW

**Comment by JoshRosen** [comment]
While I realize that this has the potential to create merge conflicts, it would be super helpful if you could squash all of the commits which came from #10835 into a single commit here in order to make it easier to spot the delta from that PR to this one.

File: core/src/main/scala/org/apache/spark/InternalAccumulator.scala [hunk 1]
```diff
@@ -17,42 +17,193 @@
...
+  }
+
+  /**
+   * Accumulators for tracking output metrics.
+   */
+  def createOutputAccums(): Seq[Accumulator[_]] = {
+    Seq[String](
+      output.WRITE_METHOD,
+      output.BYTES_WRITTEN,
+      output.RECORDS_WRITTEN).map(create)
```
Comment (davies) [inline core/src/main/scala/org/apache/spark/InternalAccumulator.scala]: These will count on failures, not consistent than SQL ones.

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/InternalAccumulator.scala]: ok, I'll add a comment.

**Comment by andrewor14** [comment]
I just merged this into #10835

**Pull request closed without merging**